### ISSUE TITLE
Error handling for user defined exceptions and error codes

### ISSUE THREAD

**Issue by arturas-venskus-flf** [STEllAR-GROUP/hpx#3307]
This is the feature request for user exceptions and error codes in the context of hpx and error handling documentation improvement

## Steps to Reproduce the Problem
Application employs try/catch exception handling such as (pseudocode):
```
// app exception type
class appEceptionType1 : std::exception {};
// functor that may throw exception
class appFunctor(){ void operator()(){ throw appEceptionType1(""); } };
// function that invokes parallel/distributed hpx constructs with supplied functors
void performAlgorithmA() {  hpx::for_each( appFunctor(), ... )  }

// try/catch block handling exceptions
try{performAlgorithmA();}
catch(const appEceptionType1 const& e){cureAlgortihmA();}
catch(...){terminateApp()}
```
When invoked function performAlgorithmA  may throw few exceptions that are stored in hpx::exception_list exception type and rethrows hpx::exception_list instead of original exception type (appEceptionType1), that causes terminateApp function invokation instead of  expected call to cureAlgortihmA in try/catch block. This behaviour can be highlighted in hpx documentation

It is possible to inherit application exception from hpx::exception: class appEceptionType1 : hpx::exception {}; but this doesn't integrates well with hpx environment because there is only one macro that allows throwing standard hpx exception HPX_THROW_STD_EXCEPTION, but not inherited one. If HPX_THROW_STD_EXCEPTION is invoked with downcasted exception the hpx:exception type is rethrown instead of appEceptionType1. If the conventional throw is invoked: throw appEceptionType1, then no nice additional information is attached to exception (stacktrace, env etc.). 

Macro HPX_THROW_EXCEPTION allows to throw exceptions from error codes, but there is no possibility to add user codes for example appEceptionType1_error_code = last_hpx_err_code +1. 

In summary:
1) It would be nice to have possibility to register and throw exceptions that are inherited from hpx::exception in hpx'y way as implemented in hpx::detail::throw_exception(application_exception_type, ...) and/or 
2) it would be nice to append external user error codes to hpx ones to employ HPX_THROW_EXCEPTION (user_error_code, ...)
3) More detailed description of exception handling in hpx documentation is needed to highlight hpx framework integration with user application peculiarities

## Specifications

  - HPX Version: 1.1
  - Platform (compiler, OS): Ubuntu 16.04

Best Regards,
Arturas Venskus
